Allow the reuse of security groups

chrislovecnm opened the pull request:
Closes: https://github.com/kubernetes/kops/issues/3124

This implements the functionality to re-use shared security groups.  Often security departments will not allow a user to create specific security groups, and this allows for shared / re-used groups.

Update - I removed Update - I removed https://github.com/chrislovecnm/kops/commit/7688be0ec6338a6420702d6d433a8391e61b916d

We have dependencies between security groups.   I have code that allows for the dependant group to be modified, which is basically the modification of shared security groups.  When that happens the cluster cannot be deleted.  I am really leaning toward no having that complication.

Instead of modifying other security groups, I recommend we require the security groups.  We would remove the above SHA and implement validation

1. You always need master and node groups - if you use shared security groups
2. If you have bastion, you need API bastion group

TODO 

- [x] test with ELB and API
- [x] add validation code
- [x] add documentation

This is WIP, but the code is stable and ready for review.

chrislovecnm commented:
/assign @geojaz @gambol99 @KashifSaadat

I know that all of you have use cases for this feature.

chrislovecnm commented:
@dcowden if you want something to test :)

chrislovecnm commented:
@justinsb if you can recommend any way to make this simpler I am all ears ...

dcowden commented:
@chrislovecnm awesome! next time we build a cluster, we'll grab this version and try it out, and i'll report back.

justinsb commented on pkg/apis/kops/v1alpha1/zz_generated.conversion.go:
```diff
@@ -456,6 +458,60 @@ func Convert_kops_CloudConfiguration_To_v1alpha1_CloudConfiguration(in *kops.Clo
 	return autoConvert_kops_CloudConfiguration_To_v1alpha1_CloudConfiguration(in, out, s)
 }
 
+func autoConvert_v1alpha1_CloudControllerManagerConfig_To_kops_CloudControllerManagerConfig(in *CloudControllerManagerConfig, out *kops.CloudControllerManagerConfig, s conversion.Scope) error {
```
Separate PR for moving these functions?

justinsb commented on pkg/model/awsmodel/api_loadbalancer.go:
```diff
@@ -140,60 +140,80 @@ func (b *APILoadBalancerBuilder) Build(c *fi.ModelBuilderContext) error {
+	createSecurityGroup := true
+	if b.Cluster.Spec.SecurityGroups != nil {
+		if b.Cluster.Spec.SecurityGroups.ApiELB != nil {
+			glog.V(8).Infof("re-using security group: %s for api elb", fi.StringValue(b.Cluster.Spec.SecurityGroups.ApiELB))
```
You just set the ID field on the task; it'll then pick up the existing security group: https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/awstasks/securitygroup.go#L108-L116

chrislovecnm replied:
Need more context, because in my mind I am doing that :)

justinsb commented on cmd/kops/integration_test.go:
```diff
@@ -130,6 +130,11 @@ func TestSharedVPC(t *testing.T) {
 
+// TestSharedSecurityGroups runs the test on a configuration with a shared VPC
+func TestSharedSecurityGroups(t *testing.T) {
+	runTestAWS(t, "sec-groups.example.com", "../../tests/integration/shared_security_groups", "v1alpha2", false, 1)
```
Might be easier to just add it into the 'complex' test ... not sure

chrislovecnm replied:
less sprawl is good

chrislovecnm replied:
Well, actually that would remove all the security group stuff in the complex example. So I would rather have this test separate and not included. I can do it either way. On the fence a bit. Have you a preference?

justinsb commented on pkg/apis/kops/cluster.go:
```diff
@@ -148,11 +148,31 @@ type ClusterSpec struct {
+	BastionELB *string `json:"bastionELB,omitempty"`
+
+	// ApiELB is the name of the sg to use for the api elb
+	ApiELB *string `json:"apiELB,omitempty"`
```
Wondering if this should be on the API LoadBalancerAccessSpec .  I think probably yes.  Not sure about whether we would then apply the same reasoning to the other fields here...

chrislovecnm replied:
It does make sense to me as well.  I was trying for grouping and simplicity.

We have lots of options, but I was thinking putting all of the base security groups in the same place would make more sense.  I would literally have to document two places instead of one.  That was my approach.

You want me to move?  Happy to.

chrislovecnm replied:
Going to move ... otherwise, it will be confusing.

justinsb commented:
So can we use the existing ID mechanism on the security group task?

And, what is the use case here?  If we had the shared-network tag, would that work?

chrislovecnm commented:
> So can we use the existing ID mechanism on the security group task?

More context here?

> And, what is the use case here? If we had the shared-network tag, would that work?

Use case for the entire PR?  I am using in a shared vpc, shared-network more context please

chrislovecnm commented:
I have merged in the bazel stuff, and I need to back it out

chrislovecnm commented:
@justinsb talking with some users, I modified the security group, to a list of security groups.  More flexibility.  I was able to refactor and hopefully make the code easier to review, and removed some unneeded code.

@geojaz bump :)

chrislovecnm commented:
@geojaz / @KashifSaadat this is ready for review.  I can add more validation checking the sg groups name is the correct format, but we can iterate.

KashifSaadat commented on docs/cluster_spec.md:
```diff
@@ -335,3 +345,44 @@ Providing the name of a key already in AWS is an alternative to `--ssh-public-ke
+    loadBalancer:
+      type: Public
+      # API ELB loadbalancer list of security group id exists in the loadBalancer section
+      securityGroup: 
```
`securityGroups`

KashifSaadat commented on docs/cluster_spec.md:
```diff
@@ -335,3 +345,44 @@ Providing the name of a key already in AWS is an alternative to `--ssh-public-ke
+  # sg ids
+  securityGroups:
+    # always required
+    master: 
```
`masterGroups`

KashifSaadat commented on docs/cluster_spec.md:
```diff
@@ -335,3 +345,44 @@ Providing the name of a key already in AWS is an alternative to `--ssh-public-ke
+    master: 
+    - sg-df180cab 
+    # always required
+    node: 
```
`nodeGroups`

KashifSaadat commented on docs/cluster_spec.md:
```diff
@@ -335,3 +345,44 @@ Providing the name of a key already in AWS is an alternative to `--ssh-public-ke
+    node: 
+    - sg-2c1e0a5q 
+    # required if using a bastion with private topology
+    bastion: 
```
`bastionGroups`

KashifSaadat commented on docs/cluster_spec.md:
```diff
@@ -335,3 +345,44 @@ Providing the name of a key already in AWS is an alternative to `--ssh-public-ke
+    bastion: 
+    - sg-a3110542 
+    # required if using a bastion with private topology
+    bastionELB: 
```
`bastionELBGroups`

KashifSaadat commented on pkg/apis/kops/cluster.go:
```diff
@@ -154,6 +154,25 @@ type ClusterSpec struct {
+	// BastionGroups is the name of the sg(s) to use for the bastion
+	BastionGroups []string `json:"bastionGroups,omitempty"`
+
+	// BastionELBGroups is the name of the sg(s) to use for the bastion elb
```
Do you mean `list of the sg id(s)`?

chrislovecnm replied:
They are a list of sg ids

chrislovecnm replied:
Would that read better?

KashifSaadat commented on pkg/apis/kops/validation/validation.go:
```diff
@@ -244,3 +254,70 @@ func validateNetworkingFlannel(v *kops.FlannelNetworkingSpec, fldPath *field.Pat
+
+	// securityGroups are defined and no load balancer security groups do not exist
+	if securityGroups != nil && noLoadBalancerGroups {
+		allErrs = append(allErrs, field.Required(fldPath.Child("SecurityGroups"), "SecurityGroups must be defined"))
```
Can you confirm you correctly fall into this case when master/node SGs are defined but LB SGs are not? I tried to test it but ended up with a task error (although my setup may have been incorrect):
`error building tasks: unexpected error resolving task "SecurityGroupRule/https-elb-to-master": Unable to find task "SecurityGroup/masters.<cluster-name>", referenced from SecurityGroupRule/https-elb-to-master:.SecurityGroup`

chrislovecnm replied:
That is another bug that is fixed in the latest push.  But I do not thing validation is working fully :(

KashifSaadat commented:
Hey, just been giving this a little test. When I defined the SG lists for `masterGroups`, `nodeGroups` and `securityGroups` (for the ELB) and attempted to update the cluster, I was getting the following error (per SG):
`executor.go:109] error running task "SecurityGroup/sg-b1faefc9" (9m59s remaining to succeed): Field cannot be changed: Name`

Does the Name tag of the Security Group need to match a specific format?

chrislovecnm commented:
@KashifSaadat I think you caught the code in between and updated that @justinsb had me do. Let me test again and I will ping u. I will shoot you over the tf I am using for my test setup as well.

chrislovecnm commented:
Good news is kops is validating the security group rules:

```console
  SecurityGroupRule/api-elb-egress
  	SecurityGroup       	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-f46a7487
  	CIDR                	0.0.0.0/0
  	Egress              	true
  	Shared              	true

  SecurityGroupRule/https-api-elb-0.0.0.0/0
  	SecurityGroup       	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-f46a7487
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	443
  	ToPort              	443
  	Shared              	true

  SecurityGroupRule/https-elb-to-master
  	SecurityGroup       	name:masters.byo-vpc-public-ha.aws.k8spro.com id:sg-bb6d73c8
  	Protocol            	tcp
  	FromPort            	443
  	ToPort              	443
  	SourceGroup         	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-f46a7487
  	Shared              	true

  SecurityGroupRule/ssh-external-to-master-0.0.0.0/0
  	SecurityGroup       	name:masters.byo-vpc-public-ha.aws.k8spro.com id:sg-bb6d73c8
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	22
  	ToPort              	22
  	Shared              	true

  SecurityGroupRule/ssh-external-to-node-0.0.0.0/0
  	SecurityGroup       	name:nodes.byo-vpc-public-ha.aws.k8spro.com id:sg-fb6a7488
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	22
  	ToPort              	22
  	Shared              	true

Will modify resources:
  SecurityGroupRule/all-master-to-master
  	Shared              	 <nil> -> true

  SecurityGroupRule/all-master-to-node
  	Shared              	 <nil> -> true

  SecurityGroupRule/all-node-to-node
  	Shared              	 <nil> -> true

  SecurityGroupRule/master-egress
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-egress
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-tcp-1-4000
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-tcp-4003-65535
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-udp-1-65535
  	Shared              	 <nil> -> true
```

My TF does not have the correct rules, and kops is picking that up.  Not sure how validation is going to work ...

chrislovecnm commented:
@justinsb I do not think this is acting the way you want it to act.  With the SecGroupRules marked as shared it is not trying to create the rule or stopping the cluster from creating.  I am guessing that is because the Rule is shared?? Not certain.  If I make the rule not shared then we will potentially update rules, and cause a state where the cluster will not delete.  Introduce dependancies between shared and not shared security groups.  AKA a ton of fun.

@KashifSaadat my validation does not seem to be working, I know @justinsb is super busy, not what your knowledge level of the API validation is.

chrislovecnm commented:
So doing some testing:

I removed shared in the ELB groups, and you will notice that it will attempt to update the security group rule.  Is this how we want it to behave?  We need to implement phases to work only with security groups then.

@justinsb / @KashifSaadat ideas?

```console
$ kops update cluster byo-vpc-public-ha.aws.k8spro.com
I1009 13:38:20.672515   50460 executor.go:91] Tasks: 0 done / 82 total; 41 can run
I1009 13:38:21.942271   50460 executor.go:91] Tasks: 41 done / 82 total; 15 can run
I1009 13:38:22.276686   50460 executor.go:91] Tasks: 56 done / 82 total; 18 can run
I1009 13:38:23.276576   50460 executor.go:91] Tasks: 74 done / 82 total; 5 can run
I1009 13:38:23.366748   50460 dnsname.go:111] AliasTarget for "api.byo-vpc-public-ha.aws.k8spro.com." is "api-byo-vpc-public-ha-aws-q273ne-1698768850.us-east-1.elb.amazonaws.com."
I1009 13:38:23.534708   50460 executor.go:91] Tasks: 79 done / 82 total; 3 can run
I1009 13:38:23.765494   50460 executor.go:91] Tasks: 82 done / 82 total; 0 can run
Will create resources:
  MirrorKeystore/mirror-keystore
  	MirrorPath          	{}

  MirrorSecrets/mirror-secrets
  	MirrorPath          	{}

  SecurityGroupRule/api-elb-egress
  	SecurityGroup       	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-5dfb0d2f
  	CIDR                	0.0.0.0/0
  	Egress              	true
  	Shared              	false

  SecurityGroupRule/https-api-elb-0.0.0.0/0
  	SecurityGroup       	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-5dfb0d2f
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	443
  	ToPort              	443
  	Shared              	false

  SecurityGroupRule/https-elb-to-master
  	SecurityGroup       	name:masters.byo-vpc-public-ha.aws.k8spro.com id:sg-70ff0902
  	Protocol            	tcp
  	FromPort            	443
  	ToPort              	443
  	SourceGroup         	name:api-elb.byo-vpc-public-ha.aws.k8spro.com id:sg-5dfb0d2f
  	Shared              	false

  SecurityGroupRule/ssh-external-to-master-0.0.0.0/0
  	SecurityGroup       	name:masters.byo-vpc-public-ha.aws.k8spro.com id:sg-70ff0902
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	22
  	ToPort              	22
  	Shared              	true

  SecurityGroupRule/ssh-external-to-node-0.0.0.0/0
  	SecurityGroup       	name:nodes.byo-vpc-public-ha.aws.k8spro.com id:sg-def701ac
  	CIDR                	0.0.0.0/0
  	Protocol            	tcp
  	FromPort            	22
  	ToPort              	22
  	Shared              	true

Will modify resources:
  SecurityGroupRule/all-master-to-master
  	Shared              	 <nil> -> true

  SecurityGroupRule/all-master-to-node
  	Shared              	 <nil> -> true

  SecurityGroupRule/all-node-to-node
  	Shared              	 <nil> -> true

  SecurityGroupRule/master-egress
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-egress
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-tcp-1-4000
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-tcp-4003-65535
  	Shared              	 <nil> -> true

  SecurityGroupRule/node-to-master-udp-1-65535
  	Shared              	 <nil> -> true

Must specify --yes to apply changes

# clove @ clove-mbp in ~/Workspace/src/github.com/vorstella/devops-clipboard on git:master x [13:38:23]
$ kops delete cluster byo-vpc-public-ha.aws.k8spro.com
TYPE			NAME												ID
autoscaling-group	master-us-east-1b.masters.byo-vpc-public-ha.aws.k8spro.com					master-us-east-1b.masters.byo-vpc-public-ha.aws.k8spro.com
autoscaling-group	master-us-east-1c.masters.byo-vpc-public-ha.aws.k8spro.com					master-us-east-1c.masters.byo-vpc-public-ha.aws.k8spro.com
autoscaling-group	master-us-east-1d.masters.byo-vpc-public-ha.aws.k8spro.com					master-us-east-1d.masters.byo-vpc-public-ha.aws.k8spro.com
autoscaling-group	nodes.byo-vpc-public-ha.aws.k8spro.com								nodes.byo-vpc-public-ha.aws.k8spro.com
dhcp-options		byo-vpc-public-ha.aws.k8spro.com								dopt-ab2185d2
iam-instance-profile	masters.byo-vpc-public-ha.aws.k8spro.com							masters.byo-vpc-public-ha.aws.k8spro.com
iam-instance-profile	nodes.byo-vpc-public-ha.aws.k8spro.com								nodes.byo-vpc-public-ha.aws.k8spro.com
iam-role		masters.byo-vpc-public-ha.aws.k8spro.com							masters.byo-vpc-public-ha.aws.k8spro.com
iam-role		nodes.byo-vpc-public-ha.aws.k8spro.com								nodes.byo-vpc-public-ha.aws.k8spro.com
instance		master-us-east-1b.masters.byo-vpc-public-ha.aws.k8spro.com					i-0876fa06c1b982f21
instance		master-us-east-1c.masters.byo-vpc-public-ha.aws.k8spro.com					i-03ee9856af4c0387e
instance		master-us-east-1d.masters.byo-vpc-public-ha.aws.k8spro.com					i-080e045c76897a5c2
instance		nodes.byo-vpc-public-ha.aws.k8spro.com								i-05ac08d9dc7105521
internet-gateway	byo-vpc-public-ha.aws.k8spro.com								igw-bf2df9c6
keypair			kubernetes.byo-vpc-public-ha.aws.k8spro.com-8b:82:20:c2:d3:ed:b0:1e:a2:5a:ce:47:28:b0:87:67	kubernetes.byo-vpc-public-ha.aws.k8spro.com-8b:82:20:c2:d3:ed:b0:1e:a2:5a:ce:47:28:b0:87:67
load-balancer		api.byo-vpc-public-ha.aws.k8spro.com								api-byo-vpc-public-ha-aws-q273ne
route-table		byo-vpc-public-ha.aws.k8spro.com								rtb-a9bc73d3
route53-record		api.byo-vpc-public-ha.aws.k8spro.com.								Z151KI3YMRFBLY/api.byo-vpc-public-ha.aws.k8spro.com.
route53-record		api.internal.byo-vpc-public-ha.aws.k8spro.com.							Z151KI3YMRFBLY/api.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-b.internal.byo-vpc-public-ha.aws.k8spro.com.						Z151KI3YMRFBLY/etcd-b.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-c.internal.byo-vpc-public-ha.aws.k8spro.com.						Z151KI3YMRFBLY/etcd-c.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-d.internal.byo-vpc-public-ha.aws.k8spro.com.						Z151KI3YMRFBLY/etcd-d.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-events-b.internal.byo-vpc-public-ha.aws.k8spro.com.					Z151KI3YMRFBLY/etcd-events-b.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-events-c.internal.byo-vpc-public-ha.aws.k8spro.com.					Z151KI3YMRFBLY/etcd-events-c.internal.byo-vpc-public-ha.aws.k8spro.com.
route53-record		etcd-events-d.internal.byo-vpc-public-ha.aws.k8spro.com.					Z151KI3YMRFBLY/etcd-events-d.internal.byo-vpc-public-ha.aws.k8spro.com.
volume			b.etcd-events.byo-vpc-public-ha.aws.k8spro.com							vol-06b601842e67ad3f6
volume			b.etcd-main.byo-vpc-public-ha.aws.k8spro.com							vol-0723bfa5ba7c7f9bc
volume			c.etcd-events.byo-vpc-public-ha.aws.k8spro.com							vol-0b5c5274cf7877b8c
volume			c.etcd-main.byo-vpc-public-ha.aws.k8spro.com							vol-0a851e855cb712191
volume			d.etcd-events.byo-vpc-public-ha.aws.k8spro.com							vol-09edb085079320056
volume			d.etcd-main.byo-vpc-public-ha.aws.k8spro.com							vol-0368694cab452750d
volume			master-us-east-1b.masters.byo-vpc-public-ha.aws.k8spro.com					vol-0ba8a7f37088491dc
volume			nodes.byo-vpc-public-ha.aws.k8spro.com								vol-09d1f534e308ac171

Must specify --yes to delete cluster
```

chrislovecnm commented:
So based on dicussions with @justinsb this is going to change some :)  I am going send a PR in with the API.

chrislovecnm commented:
Closing as we have https://github.com/kubernetes/kops/pull/3625

justinsb commented:
Did you mean to close this @chrislovecnm ?
